packages/local_auth/local_auth_ios/pigeons/messages.dart

```diff
@@ -0,0 +1,113 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:pigeon/pigeon.dart';
+
+@ConfigurePigeon(PigeonOptions(
+  dartOut: 'lib/src/messages.g.dart',
+  objcHeaderOut: 'ios/Classes/messages.g.h',
+  objcSourceOut: 'ios/Classes/messages.g.m',
+  objcOptions: ObjcOptions(
+    prefix: 'FLA',
+  ),
+  copyrightHeader: 'pigeons/copyright.txt',
+))
+
+/// Pigeon version of IOSAuthMessages, plus the authorization reason.
+///
+/// See auth_messages_ios.dart for details.
+class AuthStrings {
+  /// Constructs a new instance.
+  const AuthStrings({
+    required this.reason,
+    required this.lockOut,
+    required this.goToSettingsButton,
+    required this.goToSettingsDescription,
+    required this.cancelButton,
+    required this.localizedFallbackTitle,
+  });
+
+  final String reason;
+  final String lockOut;
+  final String goToSettingsButton;
+  final String goToSettingsDescription;
+  final String cancelButton;
+  final String? localizedFallbackTitle;
+}
+
+/// Possible outcomes of an authentication attempt.
+enum AuthResult {
+  /// The user authenticated successfully.
+  success,
+
+  /// The user failed to successfully authenticate.
+  failure,
+
+  /// The authentication system was not available.
+  errorNotAvailable,
+
+  /// No biometrics are enrolled.
+  errorNotEnrolled,
+
+  /// No passcode is set.
+  errorPasscodeNotSet,
+}
+
+class AuthOptions {
+  AuthOptions(
+      {required this.biometricOnly,
+      required this.sticky,
+      required this.useErrorDialgs});
+  final bool biometricOnly;
+  final bool sticky;
+  final bool useErrorDialgs;
+}
+
+class AuthResultDetails {
+  AuthResultDetails(
+      {required this.value, this.errorMessage, this.errorDetails});
```

Comment:
is it standard/recommended to wrap error messages/details in a pigeon method call's return type? this seems a bit tedious.

Reply:
I could be missing something, but the signature of completion takes an error, so I assume the "standard" approach is to pass in the error when error occurs:  
```
completion:(void (^)(FLAAuthResultDetails *_Nullable,
                                          FlutterError *_Nullable))completion
```

About error translation, can we catch the error in `local_auth_ios.dart` and translate it, before it escapes into client logic? 

In the current approach, however, plugin authors will have to manually include error type/message/details in their custom types, rather than directly using the FlutterError type. This feels a bit tedious to do.

Reply:
> I could be missing something, but the signature of completion takes an error, so I assume the "standard" approach is to pass in the error when error occurs:
> 
> ```
> completion:(void (^)(FLAAuthResultDetails *_Nullable,
>                                           FlutterError *_Nullable))completion
> ```

The signature takes an error because that gives Pigeon feature parity with the underlying channel API, such that someone can convert existing implementations to platform channels with minimal changes if they so desire. The existence of the possibility to do something in an API doesn't mean it's always the right way to do everything.

> About error translation, can we catch the error in `local_auth_ios.dart` and translate it, before it escapes into client logic?

We could, yes. I explained why I don't want to do that in the part of the PR description that I referred to in my previous comment. Can you explain what about the argument there you disagree with?

> In the current approach, however, plugin authors will have to manually include error type/message/details in their custom types, rather than directly using the FlutterError type.

I'm not sure who "plugin authors" is. This is one plugin, which we own. I'm not forcing the entire ecosystem to use this approach if they don't want to.

Also, I've already explained that that's not the case. `details` is only there because I don't want this to be a breaking change if clients are relying on exact platform exception details (due to us not having structured errors). That may or may not be true of `message`; we'll need to evaluate whether it's useful to preserve as-is in the combined error when the plugin is reworked. I think you are over-indexing on this one particular case that's not representative of the general design. Compare to `local_auth_android`, for instance, where we just return an enum because there are no messages or details provided by the OS.

If you really want me to, I can revert that part and go back to throwing `PlatformExceptions` directly from native code like we were before, and leave a TODO to recreate this work when we standardize errors. I was trying to do more of that work now and keep iOS more consistent with Android, but it's not required.

> This feels a bit tedious to do.

Is it more tedious than defining string constants in both Dart and ObjC for every type that we want to catch and convert, manually verifying that the strings match exactly, and manually verifying that everything we throw is indeed caught in a different file written in a different language?

Reply:
> The signature takes an error because that gives Pigeon feature parity with the underlying channel API, such that someone can convert existing implementations to platform channels with minimal changes if they so desire. 

I did not know that this `error` argument is intended for parity with platform channels. Just from the signature alone, I'd expect it to be used whenever an error happens. 

> We could, yes. I explained why I don't want to do that in the part of the PR description that I referred to in my previous comment. Can you explain what about the argument there you disagree with?

I probably misinterpreted the PR description: it says throwing error, rather than passing the error to completion block. Though you are probably referring to passing the error to completion, which will trigger dart's exception to be thrown. 

> I'm not sure who "plugin authors" is. This is one plugin, which we own. I'm not forcing the entire ecosystem to use this approach if they don't want to.

I think whatever we do in 1p plugins likely will become something that other plugin authors consider as the "golden standard", so just wanna make sure it's the best practice. 

> Is it more tedious than defining string constants in both Dart and ObjC for every type that we want to catch and convert, manually verifying that the strings match exactly, and manually verifying that everything we throw is indeed caught in a different file written in a different language?

Good point. This approach is definitely less error-prune than using `FlutterError`/`PlatformException`. I think extra fields in custom types is worth the benefit then. Are you seeing a world where we don't use `FlutterError/PlatformException` at all in 1P plugins with pigeon?

Reply:
> Are you seeing a world where we don't use `FlutterError/PlatformException` at all in 1P plugins with pigeon? (i.e. the `completion`'s error argument is never used)

I don't think so at the moment; we'll need to see how things go as we apply the current guidance to more plugins. What it currently says is:
>> This means that in general, clients of a plugin should not be expected to see raw PlatformExceptions created from error responses in native code. (This is not a strict rule; failure cases that are so obscure that clients would be unlikely to actually have specific handlers for them don't necessarily need to be converted to a common exception type.)

An example of a case where I think this applies would be `url_launcher`:
- On iOS, it can throw if you pass an invalid URL according to `NSURL`'s definition; this can only happen if you either a) use the `String`-based API that we explicitly say in dangerous and you should almost never use, or b) manage to need a URL that is a valid Dart `Uri` but not a valid `NSURL`, which should be *exceedingly* rare (this is why the String-based API still exists, but I know of only one such case currently and it's very niche).
- On Android it can throw if you have no current `Activity` when you try to launch a URL, which would be a pretty weird thing to have happen in actual usage.

I think these both fall under the "unlikely to actually have specific handlers", and there's no cross-platform conceptual thing to map them to because they are so specific, so they seem like valid cases for native errors.